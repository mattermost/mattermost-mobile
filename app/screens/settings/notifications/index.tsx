// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {withDatabase, withObservables} from '@nozbe/watermelondb/react';
import {of as of$} from 'rxjs';
import {switchMap} from 'rxjs/operators';

import {Preferences} from '@constants';
import {getPreferenceValue} from '@helpers/api/preference';
import {queryPreferencesByCategoryAndName} from '@queries/servers/preference';
import {observeConfigBooleanValue, observeConfigValue} from '@queries/servers/system';
import {observeIsCRTEnabled} from '@queries/servers/thread';
import {observeCurrentUser} from '@queries/servers/user';

import NotificationSettings from './notifications';

import type {WithDatabaseArgs} from '@typings/database/database';

const enhanced = withObservables([], ({database}: WithDatabaseArgs) => {
    return {
        currentUser: observeCurrentUser(database),
        isCRTEnabled: observeIsCRTEnabled(database),
        enableAutoResponder: observeConfigBooleanValue(database, 'ExperimentalEnableAutomaticReplies'),
        enableEmailBatching: observeConfigBooleanValue(database, 'EnableEmailBatching'),
        emailInterval: queryPreferencesByCategoryAndName(database, Preferences.CATEGORIES.NOTIFICATIONS).
            observeWithColumns(['value']).pipe(
                switchMap((preferences) => of$(getPreferenceValue<string>(preferences, Preferences.CATEGORIES.NOTIFICATIONS, Preferences.EMAIL_INTERVAL, Preferences.INTERVAL_NOT_SET))),
            ),
        sendEmailNotifications: observeConfigBooleanValue(database, 'SendEmailNotifications'),
        serverVersion: observeConfigValue(database, 'Version'),
    };
});

export default withDatabase(enhanced(NotificationSettings));
