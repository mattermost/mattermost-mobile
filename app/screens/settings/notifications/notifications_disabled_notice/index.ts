// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {withDatabase, withObservables} from '@nozbe/watermelondb/react';

import NotificationsDisabledNotice from './notifications_disabled_notice';

const enhanced = withObservables([], () => {
    return {};
});

export default withDatabase(enhanced(NotificationsDisabledNotice));
