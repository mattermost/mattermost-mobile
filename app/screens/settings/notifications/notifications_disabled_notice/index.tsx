// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useCallback, useMemo} from 'react';
import {useIntl} from 'react-intl';
import {StyleSheet, View} from 'react-native';
import Permissions from 'react-native-permissions';

import SectionNotice from '@components/section_notice';
import {Screens} from '@constants';

const styles = StyleSheet.create({
    wrapper: {
        flex: 1,
        justifyContent: 'flex-start',
        marginVertical: 16,
    },
});

type NotificationsDisabledNoticeProps = {
    testID?: string;
}

const NotificationsDisabledNotice = (props: NotificationsDisabledNoticeProps) => {
    const intl = useIntl();

    const onEnableNotificationClick = useCallback(() => {
        Permissions.openSettings('notifications');
    }, []);

    const primaryButton = useMemo(() => {
        const text = intl.formatMessage({
            id: 'user_settings.notifications.notifications_disabled_notice.button',
            defaultMessage: 'Enable notifications',
        });
        return {
            onClick: onEnableNotificationClick,
            text,
            testID: 'enable-notifications-button',
        };
    }, [intl, onEnableNotificationClick]);

    return (
        <View
            testID={props.testID}
            style={styles.wrapper}
        >
            <SectionNotice
                text={intl.formatMessage({
                    id: 'user_settings.notifications.notifications_disabled_notice.body',
                    defaultMessage: 'You will still see mention badges within the app, but you will not receive push notifications on your device.',
                })}
                title={intl.formatMessage({id: 'user_settings.notifications.notifications_disabled_notice.title', defaultMessage: 'Notifications are disabled'})}
                primaryButton={primaryButton}
                type='danger'
                location={Screens.SETTINGS_NOTIFICATION_PUSH}
            />
        </View>
    );
};

export default NotificationsDisabledNotice;
