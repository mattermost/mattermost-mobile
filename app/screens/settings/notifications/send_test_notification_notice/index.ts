// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {withDatabase, withObservables} from '@nozbe/watermelondb/react';
import {of as of$} from 'rxjs';
import {switchMap} from 'rxjs/operators';

import {observeConfigValue, observeLicense} from '@queries/servers/system';

import SendTestNotificationNotice from './send_test_notification_notice';

import type {WithDatabaseArgs} from '@typings/database/database';

const enhanced = withObservables([], ({database}: WithDatabaseArgs) => {
    return {
        isCloud: observeLicense(database).pipe(switchMap((v) => of$(v?.Cloud === 'true'))),
        telemetryId: observeConfigValue(database, 'TelemetryId').pipe(switchMap((v) => of$(v || ''))),
    };
});

export default withDatabase(enhanced(SendTestNotificationNotice));
