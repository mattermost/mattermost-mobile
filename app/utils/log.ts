// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import Config from '@assets/config.json';
import keyMirror from '@utils/key_mirror';

const SentryLevels = keyMirror({debug: null, info: null, warning: null, error: null});

export function logError(...args: any[]) {
    // eslint-disable-next-line no-console
    console.error(...args);
    addBreadcrumb(SentryLevels.error, ...args);
}

export function logWarning(...args: any[]) {
    // eslint-disable-next-line no-console
    console.warn(...args);
    addBreadcrumb(SentryLevels.warning, ...args);
}

export function logInfo(...args: any[]) {
    // eslint-disable-next-line no-console
    console.log(...args);
    addBreadcrumb(SentryLevels.info, ...args);
}

export function logDebug(...args: any[]) {
    // eslint-disable-next-line no-console
    console.debug(...args);
    addBreadcrumb(SentryLevels.debug, ...args);
}

const addBreadcrumb = (logLevel: keyof typeof SentryLevels, ...args: any[]) => {
    if (Config.SentryEnabled) {
        const Sentry = require('@sentry/react-native');
        Sentry.addBreadcrumb({
            level: logLevel,
            message: args.join(','),
            type: 'console-log',
        });
    }
};
