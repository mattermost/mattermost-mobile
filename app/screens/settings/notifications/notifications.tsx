// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React, {useCallback, useMemo} from 'react';
import {defineMessages, useIntl} from 'react-intl';

import {getCallsConfig} from '@calls/state';
import SettingContainer from '@components/settings/container';
import SettingItem from '@components/settings/item';
import {General, Screens} from '@constants';
import {useServerUrl} from '@context/server';
import useAndroidHardwareBackHandler from '@hooks/android_back_handler';
import {t} from '@i18n';
import {popTopScreen} from '@screens/navigation';
import {gotoSettingsScreen} from '@screens/settings/config';
import {getEmailInterval, getEmailIntervalTexts, getNotificationProps} from '@utils/user';

import SendTestNotificationNotice from './send_test_notification_notice';

import type UserModel from '@typings/database/models/servers/user';
import type {AvailableScreens} from '@typings/screens/navigation';

const mentionTexts = defineMessages({
    crtOn: {
        id: 'notification_settings.mentions',
        defaultMessage: 'Mentions',
    },
    crtOff: {
        id: 'notification_settings.mentions_replies',
        defaultMessage: 'Mentions and Replies',
    },
    callsOn: {
        id: 'notification_settings.calls_on',
        defaultMessage: 'On',
    },
    callsOff: {
        id: 'notification_settings.calls_off',
        defaultMessage: 'Off',
    },
});

type NotificationsProps = {
    componentId: AvailableScreens;
    currentUser?: UserModel;
    emailInterval: string;
    enableAutoResponder: boolean;
    enableEmailBatching: boolean;
    isCRTEnabled: boolean;
    sendEmailNotifications: boolean;
    serverVersion: string;
}
const Notifications = ({
    componentId,
    currentUser,
    emailInterval,
    enableAutoResponder,
    enableEmailBatching,
    isCRTEnabled,
    sendEmailNotifications,
    serverVersion,
}: NotificationsProps) => {
    const intl = useIntl();
    const serverUrl = useServerUrl();
    const notifyProps = useMemo(() => getNotificationProps(currentUser), [currentUser?.notifyProps]);
    const callsRingingEnabled = useMemo(() => getCallsConfig(serverUrl).EnableRinging, [serverUrl]);

    const emailIntervalPref = useMemo(() =>
        getEmailInterval(
            sendEmailNotifications && notifyProps?.email === 'true',
            enableEmailBatching,
            parseInt(emailInterval, 10),
        ).toString(),
    [emailInterval, enableEmailBatching, notifyProps, sendEmailNotifications]);

    const goToNotificationSettingsMentions = useCallback(() => {
        const screen = Screens.SETTINGS_NOTIFICATION_MENTION;

        const id = isCRTEnabled ? t('notification_settings.mentions') : t('notification_settings.mentions_replies');
        const defaultMessage = isCRTEnabled ? 'Mentions' : 'Mentions and Replies';
        const title = intl.formatMessage({id, defaultMessage});
        gotoSettingsScreen(screen, title);
    }, [intl, isCRTEnabled]);

    const goToNotificationSettingsPush = useCallback(() => {
        const screen = Screens.SETTINGS_NOTIFICATION_PUSH;
        const title = intl.formatMessage({
            id: 'notification_settings.push_notification',
            defaultMessage: 'Push Notifications',
        });

        gotoSettingsScreen(screen, title);
    }, [intl]);

    const callsNotificationsOn = useMemo(() => Boolean(notifyProps?.calls_mobile_sound ? notifyProps.calls_mobile_sound === 'true' : notifyProps?.calls_desktop_sound === 'true'),
        [notifyProps]);
    const goToNotificationSettingsCall = useCallback(() => {
        const screen = Screens.SETTINGS_NOTIFICATION_CALL;
        const title = intl.formatMessage({
            id: 'notification_settings.call_notification',
            defaultMessage: 'Call Notifications',
        });

        gotoSettingsScreen(screen, title);
    }, [intl]);

    const goToNotificationAutoResponder = useCallback(() => {
        const screen = Screens.SETTINGS_NOTIFICATION_AUTO_RESPONDER;
        const title = intl.formatMessage({
            id: 'notification_settings.auto_responder',
            defaultMessage: 'Automatic Replies',
        });
        gotoSettingsScreen(screen, title);
    }, [intl]);

    const goToEmailSettings = useCallback(() => {
        const screen = Screens.SETTINGS_NOTIFICATION_EMAIL;
        const title = intl.formatMessage({id: 'notification_settings.email', defaultMessage: 'Email Notifications'});
        gotoSettingsScreen(screen, title);
    }, [intl]);

    const close = useCallback(() => {
        popTopScreen(componentId);
    }, [componentId]);

    useAndroidHardwareBackHandler(componentId, close);

    return (
        <SettingContainer testID='notification_settings'>
            <SettingItem
                onPress={goToNotificationSettingsMentions}
                optionName='mentions'
                label={intl.formatMessage({
                    id: isCRTEnabled ? mentionTexts.crtOn.id : mentionTexts.crtOff.id,
                    defaultMessage: isCRTEnabled ? mentionTexts.crtOn.defaultMessage : mentionTexts.crtOff.defaultMessage,
                })}
                testID='notification_settings.mentions.option'
            />
            <SettingItem
                optionName='push_notification'
                onPress={goToNotificationSettingsPush}
                testID='notification_settings.push_notifications.option'
            />
            {callsRingingEnabled &&
                <SettingItem
                    optionName='call_notification'
                    onPress={goToNotificationSettingsCall}
                    info={intl.formatMessage({
                        id: callsNotificationsOn ? mentionTexts.callsOn.id : mentionTexts.callsOff.id,
                        defaultMessage: callsNotificationsOn ? mentionTexts.callsOn.defaultMessage : mentionTexts.callsOff.defaultMessage,
                    })}
                    testID='notification_settings.call_notifications.option'
                />
            }
            <SettingItem
                optionName='email'
                onPress={goToEmailSettings}
                info={intl.formatMessage(getEmailIntervalTexts(emailIntervalPref))}
                testID='notification_settings.email_notifications.option'
            />
            {enableAutoResponder && (
                <SettingItem
                    onPress={goToNotificationAutoResponder}
                    optionName='automatic_dm_replies'
                    info={currentUser?.status === General.OUT_OF_OFFICE && notifyProps.auto_responder_active === 'true' ? 'On' : 'Off'}
                    testID='notification_settings.automatic_replies.option'
                />
            )}
            <SendTestNotificationNotice
                serverVersion={serverVersion}
                userId={currentUser?.id || ''}
            />
        </SettingContainer>
    );
};

export default Notifications;
