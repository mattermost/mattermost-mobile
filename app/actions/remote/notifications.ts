// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Platform} from 'react-native';

import {addChannelToDefaultCategory, storeCategories} from '@actions/local/category';
import {storeMyChannelsForTeam} from '@actions/local/channel';
import {storePostsForChannel} from '@actions/local/post';
import {fetchDirectChannelsInfo, fetchMyChannel, switchToChannelById} from '@actions/remote/channel';
import {fetchPostsForChannel, fetchPostThread} from '@actions/remote/post';
import {forceLogoutIfNecessary} from '@actions/remote/session';
import {fetchMyTeam} from '@actions/remote/team';
import {fetchAndSwitchToThread} from '@actions/remote/thread';
import {ActionType} from '@constants';
import DatabaseManager from '@database/manager';
import NetworkManager from '@managers/network_manager';
import {getMyChannel, getChannelById} from '@queries/servers/channel';
import {getCurrentTeamId} from '@queries/servers/system';
import {getMyTeamById, prepareMyTeams} from '@queries/servers/team';
import {getIsCRTEnabled} from '@queries/servers/thread';
import EphemeralStore from '@store/ephemeral_store';
import {logWarning} from '@utils/log';
import {emitNotificationError} from '@utils/notification';
import {processPostsFetched} from '@utils/post';

import type {Model} from '@nozbe/watermelondb';

export const fetchNotificationData = async (serverUrl: string, notification: NotificationWithData, skipEvents = false) => {
    const channelId = notification.payload?.channel_id;

    if (!channelId) {
        return {error: 'No channel Id was specified'};
    }

    try {
        const {database} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);

        const currentTeamId = await getCurrentTeamId(database);
        let teamId = notification.payload?.team_id;
        let isDirectChannel = false;

        if (!teamId) {
            // If the notification payload does not have a teamId we assume is a DM/GM
            isDirectChannel = true;
            teamId = currentTeamId;
        }

        // To make the switch faster we determine if we already have the team & channel
        const myChannel = await getMyChannel(database, channelId);
        const myTeam = await getMyTeamById(database, teamId);

        if (!myTeam) {
            const teamsReq = await fetchMyTeam(serverUrl, teamId, false);
            if (teamsReq.error || !teamsReq.memberships?.length) {
                if (!skipEvents) {
                    emitNotificationError('Team');
                }
                return {error: teamsReq.error || 'Team'};
            }
        }

        if (!myChannel) {
            // We only fetch the channel that the notification belongs to
            const channelReq = await fetchMyChannel(serverUrl, teamId, channelId);
            if (channelReq.error ||
                !channelReq.channels?.find((c) => c.id === channelId && c.delete_at === 0) ||
                !channelReq.memberships?.find((m) => m.channel_id === channelId)) {
                if (!skipEvents) {
                    emitNotificationError('Channel');
                }
                return {error: channelReq.error || 'Channel'};
            }

            if (isDirectChannel) {
                const channel = await getChannelById(database, channelId);
                if (channel) {
                    fetchDirectChannelsInfo(serverUrl, [channel]);
                }
            }
        }

        if (Platform.OS === 'android') {
            // on Android we only fetched the post data on the native side
            // when the RN context is not running, thus we need to fetch the
            // data here as well
            const isCRTEnabled = await getIsCRTEnabled(database);
            const isThreadNotification = isCRTEnabled && Boolean(notification.payload?.root_id);
            if (isThreadNotification) {
                fetchPostThread(serverUrl, notification.payload!.root_id!);
            } else {
                fetchPostsForChannel(serverUrl, channelId);
            }
        }
        return {};
    } catch (error) {
        forceLogoutIfNecessary(serverUrl, error);
        return {error};
    }
};

export const backgroundNotification = async (serverUrl: string, notification: NotificationWithData) => {
    try {
        const {database, operator} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);
        const channelId = notification.payload?.channel_id;
        let teamId = notification.payload?.team_id;
        if (!channelId) {
            throw new Error('No channel Id was specified');
        }

        if (!teamId) {
            // If the notification payload does not have a teamId we assume is a DM/GM
            const currentTeamId = await getCurrentTeamId(database);
            teamId = currentTeamId;
        }
        if (notification.payload?.data) {
            const {data, isCRTEnabled} = notification.payload;
            const {channel, myChannel, team, myTeam, posts, users, threads} = data;
            const models: Model[] = [];

            if (posts) {
                const postsData = processPostsFetched(posts);
                const isThreadNotification = isCRTEnabled && Boolean(notification.payload.root_id);
                const actionType = isThreadNotification ? ActionType.POSTS.RECEIVED_IN_THREAD : ActionType.POSTS.RECEIVED_IN_CHANNEL;

                if (team || myTeam) {
                    const teamPromises = prepareMyTeams(operator, team ? [team] : [], myTeam ? [myTeam] : []);
                    if (teamPromises.length) {
                        const teamModels = await Promise.all(teamPromises);
                        models.push(...teamModels.flat());
                    }
                }

                await storeMyChannelsForTeam(
                    serverUrl, teamId,
                    channel ? [channel] : [],
                    myChannel ? [myChannel] : [],
                    false, isCRTEnabled,
                );

                if (data.categoryChannels?.length && channel) {
                    const {models: categoryModels} = await addChannelToDefaultCategory(serverUrl, channel, true);
                    if (categoryModels?.length) {
                        models.push(...categoryModels);
                    }
                } else if (data.categories?.categories) {
                    const {models: categoryModels} = await storeCategories(serverUrl, data.categories.categories, false, true);
                    if (categoryModels?.length) {
                        models.push(...categoryModels);
                    }
                }

                await storePostsForChannel(
                    serverUrl, channelId,
                    postsData.posts, postsData.order, postsData.previousPostId ?? '',
                    actionType, users || [],
                );

                if (isThreadNotification && threads?.length) {
                    const threadModels = await operator.handleThreads({
                        threads: threads.map((t) => ({
                            ...t,
                            lastFetchedAt: Math.max(t.post.create_at, t.post.update_at, t.post.delete_at),
                        })),
                        teamId,
                        prepareRecordsOnly: true,
                    });

                    if (threadModels.length) {
                        models.push(...threadModels);
                    }
                }
            }

            if (models.length) {
                await operator.batchRecords(models, 'backgroundNotification');
            }
            return {models};
        }
        return {models: []};
    } catch (error) {
        logWarning('backgroundNotification', error);
        return {error};
    }
};

export const openNotification = async (serverUrl: string, notification: NotificationWithData) => {
    // Wait for initial launch to kick in if needed
    await new Promise((r) => setTimeout(r, 500));

    if (EphemeralStore.getProcessingNotification() === notification.identifier) {
        return {};
    }

    EphemeralStore.setNotificationTapped(true);

    const channelId = notification.payload!.channel_id!;
    const rootId = notification.payload!.root_id!;
    try {
        const {database} = DatabaseManager.getServerDatabaseAndOperator(serverUrl);

        const isCRTEnabled = await getIsCRTEnabled(database);
        const isThreadNotification = isCRTEnabled && Boolean(rootId);

        const currentTeamId = await getCurrentTeamId(database);
        const currentServerUrl = await DatabaseManager.getActiveServerUrl();
        let teamId = notification.payload?.team_id;

        if (!teamId) {
            // If the notification payload does not have a teamId we assume is a DM/GM
            teamId = currentTeamId;
        }

        if (currentServerUrl !== serverUrl) {
            await DatabaseManager.setActiveServerDatabase(serverUrl);
        }

        // To make the switch faster we determine if we already have the team & channel
        const myChannel = await getMyChannel(database, channelId);
        const myTeam = await getMyTeamById(database, teamId);

        if (myChannel && myTeam) {
            if (isThreadNotification) {
                return fetchAndSwitchToThread(serverUrl, rootId, true);
            }
            return switchToChannelById(serverUrl, channelId, teamId);
        }

        const result = await fetchNotificationData(serverUrl, notification);
        if (result.error) {
            return {error: result.error};
        }

        if (isThreadNotification) {
            return fetchAndSwitchToThread(serverUrl, rootId, true);
        }
        return switchToChannelById(serverUrl, channelId, teamId);
    } catch (error) {
        forceLogoutIfNecessary(serverUrl, error);
        return {error};
    }
};

export const sendTestNotification = async (serverUrl: string): Promise<{status?: 'OK'; error?: unknown}> => {
    try {
        const client = NetworkManager.getClient(serverUrl);
        const result = await client.sendTestNotification();
        return result;
    } catch (error) {
        forceLogoutIfNecessary(serverUrl, error);
        return {error};
    }
};
