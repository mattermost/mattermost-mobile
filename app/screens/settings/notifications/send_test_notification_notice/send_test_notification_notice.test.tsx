// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';

import {sendTestNotification} from '@actions/remote/notifications';
import {act, fireEvent, renderWithEverything} from '@test/intl-test-helper';
import TestHelper from '@test/test_helper';
import {logError} from '@utils/log';
import {tryOpenURL} from '@utils/url';

import SendTestNotificationNotice from './send_test_notification_notice';

import type Database from '@nozbe/watermelondb/Database';

const version = '10.3.0';
const oldVersion = '10.2.0';

jest.mock('@utils/url', () => {
    return {
        tryOpenURL: jest.fn(),
    };
});

jest.mock('@actions/remote/notifications', () => {
    return {
        sendTestNotification: jest.fn(),
    };
});

jest.mock('@utils/log', () => {
    return {
        logError: jest.fn(),
    };
});

const mockedSendTestNotification = jest.mocked(sendTestNotification);

const allTimerExceptSetTimeout = [
    'Date' as const,
    'hrtime' as const,
    'nextTick' as const,
    'performance' as const,
    'queueMicrotask' as const,
    'requestAnimationFrame' as const,
    'cancelAnimationFrame' as const,
    'requestIdleCallback' as const,
    'cancelIdleCallback' as const,
    'setImmediate' as const,
    'clearImmediate' as const,
    'setInterval' as const,
    'clearInterval' as const,
    'clearTimeout' as const,
];

function getBaseProps() {
    return {
        serverVersion: version,
        telemetryId: 'someId',
        userId: 'someUserId',
        isCloud: true,
    };
}

describe('SendTestNotificationNotice', () => {
    let database: Database;

    beforeAll(async () => {
        const server = await TestHelper.setupServerDatabase();
        database = server.database;
    });

    it('should match snapshot', () => {
        const wrapper = renderWithEverything(<SendTestNotificationNotice {...getBaseProps()}/>, {database});
        expect(wrapper.toJSON()).toMatchSnapshot();
    });

    it('should show nothing on older versions', () => {
        const props = getBaseProps();
        props.serverVersion = oldVersion;
        const wrapper = renderWithEverything(<SendTestNotificationNotice {...props}/>, {database});
        expect(wrapper.toJSON()).toBeNull();
    });

    it('should open the correct url for troubleshooting docs', () => {
        const wrapper = renderWithEverything(<SendTestNotificationNotice {...getBaseProps()}/>, {database});
        fireEvent.press(wrapper.getByText('Troubleshooting docs'));
        expect(tryOpenURL).toHaveBeenCalledWith('https://mattermost.com/pl/troubleshoot-notifications?utm_source=mattermost&utm_medium=in-product-cloud&utm_content=&uid=someUserId&sid=someId');
    });

    it('should call send notification action when the send notification button is clicked, and all states go through correctly', async () => {
        jest.useFakeTimers({doNotFake: allTimerExceptSetTimeout});
        const wrapper = renderWithEverything(<SendTestNotificationNotice {...getBaseProps()}/>, {database});

        let resolve: ((value: any) => void) | undefined;
        mockedSendTestNotification.mockReturnValueOnce(new Promise((r) => {
            resolve = r;
        }));
        fireEvent.press(wrapper.getByText('Send a test notification'));
        expect(sendTestNotification).toHaveBeenCalled();
        expect(wrapper.getByText('Sending a test notification')).toBeVisible();

        await act(() => {
            resolve?.({data: true});
        });
        expect(wrapper.getByText('Test notification sent')).toBeVisible();

        act(() => {
            jest.advanceTimersByTime(3001);
        });
        expect(wrapper.getByText('Send a test notification')).toBeVisible();
    });

    it('should call send notification action when the send notification button is clicked, and when it fails, all states go through correctly', async () => {
        jest.useFakeTimers({doNotFake: allTimerExceptSetTimeout});
        const wrapper = renderWithEverything(<SendTestNotificationNotice {...getBaseProps()}/>, {database});

        let resolve: ((value: any) => void) | undefined;
        mockedSendTestNotification.mockReturnValueOnce(new Promise((r) => {
            resolve = r;
        }));
        fireEvent.press(wrapper.getByText('Send a test notification'));
        expect(sendTestNotification).toHaveBeenCalled();
        expect(wrapper.getByText('Sending a test notification')).toBeVisible();

        await act(() => {
            resolve?.({error: 'some error'});
        });

        expect(wrapper.getByText('Error sending test notification')).toBeVisible();
        expect(logError).toHaveBeenCalledWith({error: 'some error'});

        act(() => {
            jest.advanceTimersByTime(3001);
        });
        expect(wrapper.getByText('Send a test notification')).toBeVisible();
    });
});
