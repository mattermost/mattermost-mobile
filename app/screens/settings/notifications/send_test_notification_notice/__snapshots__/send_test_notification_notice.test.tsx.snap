// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendTestNotificationNotice should match snapshot 1`] = `
<View
  style={
    {
      "flex": 1,
      "justifyContent": "flex-end",
      "margin": 16,
    }
  }
>
  <View
    style={
      [
        {
          "borderStyle": "solid",
          "borderWidth": 1,
        },
        {
          "backgroundColor": "rgba(93,137,234,0.08)",
          "borderColor": "rgba(93,137,234,0.16)",
        },
        {
          "borderRadius": 4,
        },
      ]
    }
    testID="sectionNoticeContainer"
  >
    <View
      style={
        {
          "alignItems": "flex-start",
          "flexDirection": "row",
          "gap": 12,
          "padding": 16,
        }
      }
    >
      <Icon
        name="lightbulb-outline"
        size={20}
        style={
          {
            "color": "#5d89ea",
          }
        }
        testID="sectionNoticeHeaderIcon"
      />
      <View
        style={
          {
            "flex": 1,
            "flexDirection": "column",
            "gap": 8,
          }
        }
      >
        <Text
          style={
            {
              "color": "#3f4350",
              "fontFamily": "OpenSans-SemiBold",
              "fontSize": 14,
              "fontWeight": "600",
              "lineHeight": 20,
              "margin": 0,
            }
          }
        >
          Troubleshooting notifications
        </Text>
        <View
          style={
            [
              {
                "alignItems": "flex-start",
                "flexDirection": "row",
                "flexWrap": "wrap",
              },
            ]
          }
          testID="markdown_paragraph"
        >
          <Text>
            <Text
              selectable={false}
              style={
                {
                  "color": "#3f4350",
                  "fontFamily": "OpenSans",
                  "fontSize": 14,
                  "fontWeight": "400",
                  "lineHeight": 20,
                }
              }
              testID="markdown_text"
            >
              Not receiving notifications? Start by sending a test notification to all your devices to check if theyâ€™re working as expected. If issues persist, explore ways to solve them with troubleshooting steps.
            </Text>
          </Text>
        </View>
        <View
          style={
            {
              "gap": 8,
              "marginTop": 8,
            }
          }
        >
          <View
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 2,
                },
                undefined,
                false,
              ]
            }
            testID="RNE_BUTTON_WRAPPER"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": false,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "#1c58d9",
                    "borderColor": "#2089dc",
                    "borderRadius": 4,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "padding": 8,
                    "paddingHorizontal": 20,
                    "paddingVertical": 10,
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "gap": 7,
                    }
                  }
                  testID="undefined-text-container"
                >
                  <Text
                    numberOfLines={1}
                    style={
                      {
                        "color": "#ffffff",
                        "fontFamily": "OpenSans-SemiBold",
                        "fontSize": 14,
                        "fontWeight": "600",
                        "lineHeight": 20,
                      }
                    }
                  >
                    Send a test notification
                  </Text>
                </View>
              </View>
            </View>
          </View>
          <View
            style={
              [
                {
                  "overflow": "hidden",
                },
                {
                  "borderRadius": 2,
                },
                undefined,
                false,
              ]
            }
            testID="RNE_BUTTON_WRAPPER"
          >
            <View
              accessibilityRole="button"
              accessibilityState={
                {
                  "busy": false,
                  "checked": undefined,
                  "disabled": false,
                  "expanded": undefined,
                  "selected": undefined,
                }
              }
              accessibilityValue={
                {
                  "max": undefined,
                  "min": undefined,
                  "now": undefined,
                  "text": undefined,
                }
              }
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "opacity": 1,
                }
              }
            >
              <View
                style={
                  {
                    "alignItems": "center",
                    "backgroundColor": "rgba(28,88,217,0.08)",
                    "borderColor": "#2089dc",
                    "borderRadius": 4,
                    "borderWidth": 0,
                    "flexDirection": "row",
                    "justifyContent": "center",
                    "padding": 8,
                    "paddingHorizontal": 20,
                    "paddingVertical": 10,
                  }
                }
              >
                <View
                  style={
                    {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "gap": 7,
                    }
                  }
                  testID="undefined-text-container"
                >
                  <View>
                    <Icon
                      color="#1c58d9"
                      name="open-in-new"
                      size={18}
                      testID="undefined-icon"
                    />
                  </View>
                  <Text
                    numberOfLines={1}
                    style={
                      {
                        "color": "#1c58d9",
                        "fontFamily": "OpenSans-SemiBold",
                        "fontSize": 14,
                        "fontWeight": "600",
                        "lineHeight": 20,
                      }
                    }
                  >
                    Troubleshooting docs
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
